editor.planx.uk/src/pages/FlowEditor/components/Sidebar/EditHistory/Timeline.tsx

```diff
@@ -0,0 +1,171 @@
+import React, { useState } from "react";
+import { FONT_WEIGHT_SEMI_BOLD } from "theme";
+import { CommentHistoryItem, HistoryItem, OperationHistoryItem, PublishHistoryItem } from ".";
+import { CommentTimelineItem, OperationTimelineItem, PublishTimelineItem } from "./TimelineItems";
+
+interface EditHistoryTimelineProps {
+  events: HistoryItem[];
+}
+
+export const EditHistoryTimeline = ({ events = [] }: EditHistoryTimelineProps) => {
+  const [focusedOpIndex, setFocusedOpIndex] = useState<number | undefined>(
+    undefined,
+  );
+
+  const [flow, teamSlug, canUserEditTeam, undoOperation, deleteFlowComment] = useStore(
+    (state) => [
+      state.flow,
+      state.teamSlug,
+      state.canUserEditTeam,
+      state.undoOperation,
+      state.deleteFlowComment,
+    ],
+  );
+
+  const handleUndo = (i: number) => {
+    // Get all events since & including the selected one
+    const eventsToUndo = events.slice(0, i + 1);
+
+    const operationsToUndo = eventsToUndo.filter((event) => event.type === "operation") as OperationHistoryItem[];
+    const commentsToDelete = eventsToUndo.filter((event) => event.type === "comment") as CommentHistoryItem[];
+
+    // Make a flattened list of operations, with the latest operations first
+    const operationsData: Array<OT.Op[]> = [];
+    operationsToUndo.map((op) => operationsData.unshift(op.data));
+    const flattenedOperationsData: OT.Op[] = operationsData?.flat(1);
+
+    // Then undo all operations by applying a single new inverse operation on top
+    undoOperation(flattenedOperationsData);
+
+    // Also delete each comment in the undo scope
+    commentsToDelete.forEach((comment) => {
+      deleteFlowComment(comment.id);
+    });
+  };
+
+  const inUndoScope = (i: number): boolean => {
+    // Is a given operation in the list in scope of also being "undone" if the currently focused button is clicked?
+    return focusedOpIndex !== undefined && i < focusedOpIndex;
+  };
+
+  const isUndoType = (type: HistoryItem["type"]): boolean => {
+    // Only operations and comments will be "undone" when restoring to a point in time, publishes are permanent snapshots
+    return ["operation", "comment"].includes(type);
+  };
+
+  const showUndoButton = (event: HistoryItem): boolean => {
+    // Only show the restore button for operations within teams I can edit, omitting the intial default operation for new flows which won't have an actor 
+    return event.type === "operation" && Boolean(event.actorId) && canUserEditTeam(teamSlug);
+  };
+
+  return (
+    <Timeline
+      sx={{
+        padding: 0,
+        [`& .${timelineItemClasses.root}:before`]: {
+          flex: 0,
+          padding: 0,
+        },
+      }}
+    >
+      {events?.map((op: HistoryItem, i: number) => (
+        <TimelineItem key={`${op.type}-${op.id}`}>
+          <TimelineSeparator>
+            <TimelineDot
+              sx={{
+                bgcolor: (theme) =>
+                  inUndoScope(i) && isUndoType(op.type)
```

Comment:
nit: We're relying on this combination in line a few places - maybe a sign that `TimelineItem` should be a component with this as a single calculated value.